### DISCUSSION TITLE
LLHLS DRM Feature Released

### DISCUSSION THREAD

**Discussion by getroot** [OvenMediaLabs/OvenMediaEngine#1375] [Announcement]
I am thrilled to announce the addition of DRM (Digital Rights Management) to OvenMediaEngine, a highly anticipated feature by many in the community. DRM support is now available through LLHLS Publisher, and you can seamlessly integrate it into your streaming workflow. Here's how you can make use of it:

To include DRM information in your LLHLS Publisher configuration, follow these steps. You can set the InfoFile path as either a relative path, starting from the directory where Server.xml is located, or as an absolute path.

```
<LLHLS>
    <ChunkDuration>0.5</ChunkDuration>
    <PartHoldBack>1.5</PartHoldBack>
    <SegmentDuration>6</SegmentDuration>
    <SegmentCount>10</SegmentCount>
    <DRM>
        <Enable>false</Enable>
        <InfoFile>path/to/file.xml</InfoFile>
    </DRM>
    <CrossDomains>
        <Url>*</Url>
    </CrossDomains>
</LLHLS>
```

The separation of the DRMInfoFile is designed to allow dynamic changes to the file. Any modifications to the DRMInfoFile will take effect when a new stream is generated.

Here's how you should structure your DRM Info File:

```
<?xml version="1.0" encoding="UTF-8"?>

<DRMInfo>
    <DRM>
        <Name>MultiDRM</Name>
        <VirtualHostName>default</VirtualHostName>
        <ApplicationName>app</ApplicationName>
        <StreamName>stream*</StreamName> <!-- Can be a wildcard regular expression -->
        <CencProtectScheme>cbcs</CencProtectScheme> <!-- Currently supports cbcs only -->
        <KeyId>572543f964e34dc68ba9ba9ef91d4xxx</KeyId> <!-- Hexadecimal -->
        <Key>16cf4232a86364b519e1982a27d90xxx</Key> <!-- Hexadecimal -->
        <Iv>572547f914e34dc68ba9ba9ef91d4xxx</Iv> <!-- Hexadecimal -->
        <Pssh>0000003f7073736800000000edef8ba979d64acea3c827dcd51d21ed0000001f1210572547f964e34dc68ba9ba9ef91d4c4a1a05657a64726d48f3c6899xxx</Pssh> <!-- Hexadecimal, for Widevine -->
        <!-- Add Pssh for FairPlay if needed -->
        <FairPlayKeyUrl>skd://fiarplay_key_url</FairPlayKeyUrl> <!-- FairPlay only -->
    </DRM>
    <DRM>
        <Name>MultiDRM2</Name>
        <VirtualHostName>default</VirtualHostName>
        <ApplicationName>app2</ApplicationName>
        <StreamName>stream*</StreamName> <!-- Can be a wildcard regular expression -->
         ...........
    </DRM>
</DRMInfo>
```

Multiple `<DRM>` can be set.  Specify the VirtualHost, Application, and StreamName where DRM should be applied. StreamName supports wildcard regular expressions.

Currently, CencProtectScheme only supports "cbcs" since FairPlay also supports only cbcs. There may be limited prospects for adding other schemes in the near future.

KeyId, Key, Iv and Pssh values are essential and should be provided by your DRM provider. FairPlayKeyUrl is only need for FairPlay and if you want to enable FairPlay to your stream, it is required. It will be also provided by your DRM provider. 

OvenPlayer now includes DRM-related options. Enable DRM and input the License URL. Your content is now securely protected.

![image](https://github.com/AirenSoft/OvenMediaEngine/assets/17779085/e8cfc94f-d362-4551-952e-812df1170551)

**Note: Currently, DRM is only supported for H.264 and AAC codecs. Support for H.265 will be added soon.**

You can test it with the current Master Branch, and it will be released in the next version after stability has been verified.

**Comment by getroot** [comment]
I've fully tested Widevine, but I haven't fully tested FairPlay yet because I don't have a server certificate for FairPlay. (Although all specifications were met)
I would greatly appreciate it if someone could test FairPlay or provide information for testing FairPlay.

**Comment by getroot** [comment]
I updated OvenMediaEngine to make it easier to integrate with Pallycon.

To integrate Pallycon, configure the DRMInfo.xml file as follows.

```
<?xml version="1.0" encoding="UTF-8"?>

<DRMInfo>
    <DRM>
        <Name>Pallycon</Name>
        <VirtualHostName>default</VirtualHostName>
        <ApplicationName>app</ApplicationName>
        <StreamName>stream*</StreamName> <!-- Can be wildcard regular expression -->

        <DRMProvider>Pallycon</DRMProvider> <!-- Manual(default), Pallycon -->
        <DRMSystem>Widevine,Fairplay</DRMSystem> <!-- Widevine, Fairplay -->
        <CencProtectScheme>cbcs</CencProtectScheme> <!-- cbcs -->
        <ContentId>${VHostName}_${AppName}_${StreamName}</ContentId>
        <KMSUrl>https://kms.pallycon.com/v2/cpix/pallycon/getKey/</KMSUrl>
        <KMSToken>xxxx</KMSToken>
    </DRM>
</DRMInfo>

```

Set `<DRMProvider>` to Pallycon, then enter other information. `ContentId `is required for data you set in the player. `KMSUrl `and `KMSToken `are provided when you sign up for Pallycon. For more information about Pallycon, please visit their website. https://pallycon.com/docs/en/multidrm/

**Comment by getroot** [comment]
I confirmed that FairPlay works well with Pallycon. (There were a few patches)

**Reply by dnetcrawler** [reply]
Do we need to update the container to get Pallycon to work? When we apply these settings it is still sending the stream in the clear.

**Reply by getroot** [reply]
You must download the latest code from the master branch or use airensoft/ovenmediaengine:dev. It will be included in the next release 0.16.4.
